stubs/SQLAlchemy/sqlalchemy/engine/base.pyi

```diff
@@ -82,88 +118,115 @@ class ExceptionContextImpl(ExceptionContext):
     ) -> None: ...
 
 class Transaction(TransactionalContext):
-    def __init__(self, connection) -> None: ...
+    def __init__(self, connection: Connection) -> None: ...
     @property
-    def is_valid(self): ...
+    def is_valid(self) -> bool: ...
     def close(self) -> None: ...
     def rollback(self) -> None: ...
     def commit(self) -> None: ...
+    # The following field are technically not defined on Transaction, but on
+    # all sub-classes.
+    @property
+    @abstractmethod
+    def connection(self) -> Connection: ...
+    @property
+    @abstractmethod
+    def is_active(self) -> bool: ...
 
 class MarkerTransaction(Transaction):
-    connection: Any
-    def __init__(self, connection) -> None: ...
+    connection: Connection
     @property
-    def is_active(self): ...
+    def is_active(self) -> bool: ...
 
 class RootTransaction(Transaction):
-    connection: Any
+    connection: Connection
     is_active: bool
-    def __init__(self, connection) -> None: ...
 
 class NestedTransaction(Transaction):
-    connection: Any
+    connection: Connection
     is_active: bool
-    def __init__(self, connection) -> None: ...
 
 class TwoPhaseTransaction(RootTransaction):
     xid: Any
-    def __init__(self, connection, xid) -> None: ...
+    def __init__(self, connection: Connection, xid) -> None: ...
     def prepare(self) -> None: ...
 
-class Engine(Connectable, log.Identified):
-    pool: Any
-    url: Any
-    dialect: Any
-    logging_name: Any
-    echo: Any
-    hide_parameters: Any
+class Engine(Connectable, Identified):
+    pool: Pool
+    url: str
+    dialect: Dialect
+    logging_name: str  # only exists if not None during initialization
+    echo: echo_property
```

Comment:
Maybe this should be `ClassVar`, because the class has an instance of `echo_property`. Or just `echo: _EchoFlag`, since the underlying `echo_property` instance is an implementation detail?

Reply:
I'm not sure how well class vars play together with the descriptor protocol. `@property` is usually not wrapped in `ClassVar` either. I just tried to use just `_EchoFlag`, and while it should work, stubtest complains, so I opted for the version that requires less stubtest overrides.